library/std/src/sys/windows/rand.rs

```diff
@@ -34,35 +33,94 @@ use crate::sys::c;
 /// [`HashMap`]: crate::collections::HashMap
 /// [`RandomState`]: crate::collections::hash_map::RandomState
 pub fn hashmap_random_keys() -> (u64, u64) {
-    let mut v = (0, 0);
-    let ret = unsafe {
-        let size = mem::size_of_val(&v).try_into().unwrap();
-        c::BCryptGenRandom(
-            // BCRYPT_RNG_ALG_HANDLE is only supported in Windows 10+.
-            // So for Windows 8.1 and Windows 7 we'll need a fallback when this fails.
-            ptr::invalid_mut(c::BCRYPT_RNG_ALG_HANDLE),
-            ptr::addr_of_mut!(v).cast(),
-            size,
-            0,
-        )
-    };
-    if ret != 0 { fallback_rng() } else { v }
+    Rng::open().and_then(|rng| rng.gen_random_keys()).unwrap_or_else(fallback_rng)
+}
+
+struct Rng(c::BCRYPT_ALG_HANDLE);
+impl Rng {
+    #[cfg(miri)]
+    fn open() -> Result<Self, c::NTSTATUS> {
+        const BCRYPT_RNG_ALG_HANDLE: c::BCRYPT_ALG_HANDLE = ptr::invalid_mut(0x81);
+        let _ = (
+            c::BCryptOpenAlgorithmProvider,
+            c::BCryptCloseAlgorithmProvider,
+            c::BCRYPT_RNG_ALGORITHM,
+            c::STATUS_NOT_SUPPORTED,
+        );
+        Ok(Self(BCRYPT_RNG_ALG_HANDLE))
+    }
```

Comment:
I am surprised to see a Miri special path here; usually we try to execute the real thing where feasible (and `BCryptOpenAlgorithmProvider` looks easy enough to fake). Also please Cc @rust-lang/miri when adding such a codepath.

Reply:
Yeah, for this I'd say it's better to just add the shims to miri.

Reply:
Sorry! My intent was for this to be temporary until I was confident that the crash issue was addressed. Then we can go back to using `BCRYPT_USE_SYSTEM_PREFERRED_RNG` as the default and `BCryptOpenAlgorithmProvider` would be the fallback. I can submit a PR to undo tho.

Reply:
Ah, if this is just temporary then implementing the shims in Miri would indeed not be worth it.